Add Supabase GoTrue Auth Cookbook

keithtelliott opened the pull request:
Hi Redwooders, 

I created a Supabase version of @cannikin's GoTrue Auth cookbook and posted it on the [forums](https://community.redwoodjs.com/t/cookbook-supabase-gotrue-auth/2531).  Rob suggested that we add it to the redwoodjs.com web site - and this PR will do just that.  

This is my first Redwood PR - what fun!

Some questions...

- Where do I update navigation links?  (what file?  I'll update the nav, but couldn't find it)
- Any writing style comments?  Feel free to be picky.
- What else?  

Thanks,
   Keith

dthyresson commented:
@keithtelliott Thanks for this. Will review. Two things:

1. Needs a menu side-bar link
2. I see you referenced the Create React App. Did you See the Redwood Quick Start?

https://supabase.com/docs/guides/with-redwoodjs

dthyresson commented on cookbook/Supabase_GoTrue_Auth.md:
```diff
@@ -0,0 +1,717 @@
+
+  const onSubmit = (data) => {
+    setError(null)
+    logIn({ email: data.email, password: data.password })
```
Have you considered `await` vs promise, then?

keithtelliott replied:
I simply followed the lead from the prior [Netlify GoTrue](https://github.com/redwoodjs/redwoodjs.com/blob/main/cookbook/GoTrue_Auth.md) cookbook.  Rob uses the promise/then style, so I rolled with it.  I'm inclined to keep it as-is, but I could be convinced otherwise...

dthyresson replied:
I would use await since that matches Supabase's API document reference.

keithtelliott commented:
Thanks for jumping in and reviewing @dthyresson.  

I did NOT see the Supabase Redwood Quick Start.  It must be hot of the press, because it was not there when I was climbing the learning curve in Oct/Nov.  

So, let me go through it and adjust the content appropriately.  This may take me a few days as I juggle chill holiday family time...

keithtelliott commented:
Also, @dthyresson would you be able to ask your Supabase friends to add a Redwood button/link on their Getting Started page?  A link does appear on their left nav, but not on the main page (and on my phone the left nav typically does not appear, so your guide is easily missed).  I set out to submit a PR for their site... but it's not on github :( 

<img width="762" alt="image" src="https://user-images.githubusercontent.com/43206213/147278754-f0336059-9091-4f79-a505-2c8d552981fa.png">

dthyresson commented:
> Also, @dthyresson would you be able to ask your Supabase friends to add a Redwood button/link on their Getting Started page? 

I was thinking the same thing. Fortunately the Studio is no open-sourced and also the web docs are: https://github.com/supabase/supabase/blob/62cf98332b4a4424a3a6cb44c1d18ceb712fe2e4/web/docs/about.mdx

So can make a PR for those.

Good idea.

dthyresson commented:
@keithtelliott I have added issue https://github.com/supabase/supabase/issues/4650 and will have a PR up shortly.

dthyresson commented:
@keithtelliott See PR https://github.com/supabase/supabase/pull/4651

keithtelliott commented:
@dthyresson, Way to go on the quick work to get the Supabase docs updated - You are fast and effective!

I updated my docs, integrating comments regarding the Supabase Redwood Quick Start.  

Note that I needed to spend time working directly with the Supabase client, to gain an understanding of their client.  After that I was able to confidently use the Redwood auth hook/wrappers. then I started to use the Redwood auth wrappers.  So, my comments describe this (I recommend working through the React Quick Start first, then the Redwood Quick Start).  

Please le me know what you think.

Also, I still need to update links/nav on this site, and will provide another commit once I do so.

dthyresson commented:
> Also, I still need to update links/nav on this site, but I'm stuck: in what files do the navigation links live?

I have added the nav link menu. It's in a `build.js` file which isn't that intuitive, but there it is.

I have a few comments that I will add in my review based on a few things I have recently learned about Supabase auth having used it more and done some PR's in GoTrue and the GoTrue js client.

dthyresson commented on cookbook/Supabase_GoTrue_Auth.md:
```diff
@@ -0,0 +1,711 @@
+const SignupPage = () => {
+  const { client } = useAuth()
+
+  const onSubmit = (data) => {
```
I know you say below about Promises, but most of the Supabase docs use `await` and think it fits better.

Also, logging the result is fine for dev, but shouldn't be done in a prod app.

I'd also suggest using Redwood's toast to show the error message.

No one will ever see a logged error in the JS console.

keithtelliott replied:
I updated the code to use async/await.  

I left the console logs in there, because they are just part of the initial example and are removed as the reader continues through the cookbook.   The final results of the cookbook do not include any console logs.

dthyresson commented on cookbook/Supabase_GoTrue_Auth.md:
```diff
@@ -0,0 +1,711 @@
+  const onSubmit = (data) => {
+    setError(null)
+    client
+      .signup(data.email, data.password)
```
Same for awaits and logging vs toast messages

dthyresson commented on cookbook/Supabase_GoTrue_Auth.md:
```diff
@@ -0,0 +1,711 @@
+  const { client } = useAuth()
+  const [error, setError] = React.useState(null)
+
+  const onSubmit = (data) => {
```
Have you seen the code samples in the auth playground here: https://github.com/redwoodjs/playground-auth/blob/main/web/src/lib/code-samples/supabase.md

I would *highly* recommend adding a section on the authentication state change listening:

```js
// Here you can subscribe to events
  client.auth.onAuthStateChange(async (event) => {
    if (event === 'SIGNED_IN') {
      console.debug('>> in onAuthStateChange', event)
    }

    if (event === 'SIGNED_OUT') {
      console.debug('>> in onAuthStateChange', event)
      // reset the auth state to ensure no longer authenticated in all tabs
      await reauthenticate()
    }

    if (event === 'TOKEN_REFRESHED') {
      console.debug('>> in onAuthStateChange', event)
    }
  })
```

This lets you listen for events and say add toast messages and also demo's that tokens refresh.,

Token refresh is a huge and important consideration. It is the JWT expiration setting in Supabase.

By default, the client will autoRefresh as seen here: https://github.com/supabase/gotrue-js/blob/d7b334a4283027c65814aa81715ffead262f0bfa/src/GoTrueClient.ts#L24

It may be worth mentioning this -- that the token will auto refresh.

keithtelliott replied:
See my note below.  I like the thought of including this, but I recommend rolling this into a second cookbook in the future.

dthyresson commented on cookbook/Supabase_GoTrue_Auth.md:
```diff
@@ -0,0 +1,711 @@
+- [Redwood Supabase Auth Installation & Setup](https://redwoodjs.com/docs/authentication#supabase)
+- [Redwood Auth Playground](https://redwood-playground-auth.netlify.app/supabase)
+- [Redwood Supabase Auth Client Implementation](https://github.com/redwoodjs/redwood/blob/main/packages/auth/src/authClients/supabase.ts)
+
```
Not sure if you want to mention:

* Passwordless auth (just email and get a magic confirmation link)
* Phone one-time-password auth
* Sign in with refresh token
* Third-party auth via GitHub etc
* Link to Auth playground

keithtelliott replied:
I’ll update the narrative to include comments on these topics.  Also, will hold onto a vision of future cookbooks to go deep on some in detail.

keithtelliott commented:
Good thorough comments.  Tomorrow I’ll study in detail and make another commit ASAP…

dthyresson commented:
@keithtelliott a nice example of the user trigger can be found here https://twitter.com/divjoy/status/1479201184065806337?s=21

keithtelliott commented:
@dthyresson I made a little progress this weekend.  I changed from promise/then to async/await.  I also renamed from Supabase GoTrue Auth to Supabase Auth.  

I propose that we drive toward creating a second future cookbook (or more) to roll in additional concepts.  Along these lines, I'll make an additional commit within a couple days that hits on some/all of your suggestions in the narrative, then leaves a teaser for a future Cookbook that will go in depth.  

Because, for noobs, these initial baby steps can be lots to consume.  So breaking into separate files can help not-overload folks.  Second, I don't have much time at the moment - I'd like to call an early chunk done, then bolt on more later.

Please let me know your thoughts - I value your insight and expertise.

keithtelliott commented:
OK, just added some more narrative commentary.  Let me know what you think.  My suggestion: wrap this baby up and plan on future cookbooks to go deeper in addition details...

dthyresson commented:
@jtoar Could I get a copy review and approve and merge when good? Thanks!

jtoar reviewed (approved):
@keithtelliott thanks for your patience—this is awesome! I made a lot of tiny edits for consistency. Let's get this in!
